File: app/src/domain/calculations/invMethod.ts
import { safeDivide } from './utils'

/**
 * 在庫法（Inventory Method）計算
 *
 * スコープ: 全売上・全仕入（花・産直含む全体）
 * 目的: 実際の在庫実績に基づく会計的な粗利を算出
 */

/** 在庫法の入力パラメータ */
export interface InvMethodInput {
  readonly openingInventory: number | null // 期首在庫
  readonly closingInventory: number | null // 期末在庫
  readonly totalPurchaseCost: number // 総仕入原価
  readonly totalSales: number // 総売上高
}

/** 在庫法の計算結果 */
export interface InvMethodResult {
  readonly cogs: number | null // 売上原価
  readonly grossProfit: number | null // 粗利益
  readonly grossProfitRate: number | null // 粗利率
}

/**
 * 在庫法による実績粗利を計算する
 *
 * 売上原価 = 期首在庫 + 総仕入高 - 期末在庫
 * 粗利益   = 売上高 - 売上原価
 * 粗利率   = 粗利益 / 売上高
 */
export function calculateInvMethod(input: InvMethodInput): InvMethodResult {
  const { openingInventory, closingInventory, totalPurchaseCost, totalSales } = input

  // 期首/期末在庫が不明な場合は計算不可
  if (openingInventory == null || closingInventory == null) {
    return { cogs: null, grossProfit: null, grossProfitRate: null }
  }

  const cogs = openingInventory + totalPurchaseCost - closingInventory
  const grossProfit = totalSales - cogs
  const grossProfitRate = safeDivide(grossProfit, totalSales, 0)

  return { cogs, grossProfit, grossProfitRate }
}

File: wasm/gross-profit/src/inv_method.rs
use crate::types::InvMethodResult;
use crate::utils::safe_divide;

/// Calculate gross profit using the inventory method.
///
/// Null propagation: if either opening or closing inventory is NaN (null sentinel),
/// the entire result is null (is_null = true).
///
/// Steps:
/// 1. cogs = openingInventory + totalPurchaseCost - closingInventory
/// 2. grossProfit = totalSales - cogs
/// 3. grossProfitRate = grossProfit / totalSales  [fallback: 0]
pub fn calculate_inv_method(
    opening_inventory: f64,
    closing_inventory: f64,
    total_purchase_cost: f64,
    total_sales: f64,
) -> InvMethodResult {
    if opening_inventory.is_nan() || closing_inventory.is_nan() {
        return InvMethodResult {
            is_null: true,
            cogs: 0.0,
            gross_profit: 0.0,
            gross_profit_rate: 0.0,
        };
    }

    let cogs = opening_inventory + total_purchase_cost - closing_inventory;
    let gross_profit = total_sales - cogs;
    let gross_profit_rate = safe_divide(gross_profit, total_sales, 0.0);

    InvMethodResult {
        is_null: false,
        cogs,
        gross_profit,
        gross_profit_rate,
    }
}
